packages/testing/src/render.js

```diff
@@ -0,0 +1,22 @@
+import React from 'react'
+import { render } from '@testing-library/react'
+import { MockedProvider } from '@apollo/react-testing'
+
+import Routes from '~/web/Routes'
+
+const AllTheProviders = ({ mocks, children }) => {
+  return (
+    <MockedProvider mocks={mocks} addTypename={false}>
+      <>
+        <Routes />
+        {children}
+      </>
+    </MockedProvider>
+  )
+}
+
+export const customRender = (ui, options = {}) =>
+  render(ui, {
+    wrapper: (props) => <AllTheProviders mocks={options.mocks} {...props} />,
+    ...options,
+  })
```

Comment:
I'm curious if we can improve the API here?

Currently a test would use:

```js
render(<BlogPostPage id="id-123" />, { mocks })
```

In a more complex situation, it might eventually look like this:

```js
render(<BlogPostPage id="id-123" />, { mocks, useTypeName: true, resolvers: localResolvers,  defaultOptions: apolloOptions })
```

While these options are less common, it may be good to still expose them (especially `useTypeName` for reasons I mentioned above). It feels like in doing so we will start overloading the `render` function too much.

What are your thoughts on exporting an additional top level component from this file that controls mocking, and exporting the custom render separately without the mocks? This would have the additional benefit of allowing unmocked Apollo queries should we need to, as well as handle additional [Apollo options](https://www.apollographql.com/docs/react/api/react-testing/#mockedprovider) more easily.

It might look something like this:

```ts
interface MockedTestStateProps {
  children: ReactNode;
  mockOptions: MockedProviderProps;
}

export const WithMockedTestState = ({ children, mockOptions = {}}: MockedTestStateProps) => {
  <MockedProvider {...mockOptions}>
    {children}
  </MockedProvider>
}
```

The example above would be written as:

```js
render(
  <WithMockedTestState mocks={mocks}>
    <BlogPostPage id="id-123" />
  </WithMockedTestState>
)
```

And for the complex case:

```js
render(
  <WithMockedTestState mocks={mocks} resolvers={localResolvers} defaultOptions={apolloOptions}>
    <BlogPostPage id="id-123" />
  </WithMockedTestState>
)
```

When Apollo mocking is either not desired, or the user is testing a component that doesn't use Apollo, it could still be written as:

```js
render(
  <BlogPostPage id="id-123" />
)
```

As a future point, we may want to consider a way to pass additional `Providers` to the custom render (like theme or other context providers) but for now, I think it's more than fine they are passed explicitly.

Reply:
I see your point. I haven't used Apollo Mocks before at all so it's good to get the perspective from someone who has!

I guess my idea was to save people from having to wrap the `MockProvider` in every test where the mock part _isn't_ complicated, which I can imagine are most tests.

I like the idea of splitting them up so it's more flexible tho. But in that case I'm wondering what is the additional benefit of the component we put in between as opposed to just importing the entire `MockProvider` in your app. Unless of course we want to add other stuff later! In that case I think it's good to just re-export `MockProvider` directly from apollo. Or maybe it's an idea to export 2 renders; one with and one without the mocks. But that might be messy/weird? What do you think? 

As for adding additional Providers, this is already possible! You can use the `render` function exposed by this PR, and just go:
```javascript
render(
  <Provider>
    <Subject />
  </Provider>
)
```

You could even extend the render function again in your application if you'd like!

Reply:
I'll let someone from the Core Team chime in on this. This is really an API design consideration since both are functionally equivalent.

To me, explicitly adding the wrapper component for mocks is clearer and more explicit than adding additional options to the React Testing Library render function. Of course, it's entirely possible that I'm the outlier in that thought. 

```js
render(
  <WithMockedTestState mocks={mocks}>
    <BlogPostPage id="id-123" />
  </WithMockedTestState>
)
```
vs

```js
render(
  <BlogPostPage id="id-123" />,
  { mocks }
)
```

Reply:
I think both approaches have merit: 

1. The `render(Component, { mocks: [] })` is nice because it's less typing, but it may be unexpected. 
2. Using a component is more explicit, but it isn't very different to importing `MockProvider` and wrapping it yourself.

I don't think we have a very compelling integration path for mocking graphql response data yet, but we'll get there. I think we should stick with the `render` argument, but we could name it something else:
```js
render(..., { sdl: { mocks: {}, resolvers: {}, ...etc } } )
```
Maybe we could conditionally render the `MockProvider` if `sdl` is present?

Some ideas for the future:
1. Have an single place to put mock data that works across all the layers of the `db -> sdl -> resolvers -> services.`
2. Figure out a way to generate mock data based off "field names". As an example we could easily generate somewhat realistic data for the following: name, surname, address, title, slug, email, and for the fields that cannot be determined we could figure something out.

Reply:
It took me a bit, but I think I understand where you are going with the `sdl` and mocking the full api or just `resolvers` or even other layers.

I do think it's quite a big API decision like @cball mentioned. 

How would you feel about taking the `MockedProvider` out of the custom `render` method for now, writing a small recipe/guide/doc for current users on how to import `MockedProvider` and then we can discuss a bit more on what the API should look like...

OR

We leave it in for now, perhaps in the format you suggested, we see what people think and change it later... Redwood is in early stages after all.